### DISCUSSION TITLE
How to chat inside the VLC interface like on MPV

### DISCUSSION THREAD

**Discussion by jvschiavo** [Syncplay/syncplay#470] [Archive]
Hi, I'd like to know if it is possible to chat inside VLC like on mpv by pressing enter. I have not managed to be able to do that, and didn't find any information regarding this.

(side-note, chatting on mpv on Linux only lets me use ASCII characters and not the special characters of portuguese like ã, ó, etc, but I can do so inside the syncplay window, does anyone know why?)

**Comment by Et0h** [comment]
> Hi, I'd like to know if it is possible to chat inside VLC like on mpv by pressing enter. I have not managed to be able to do that, and didn't find any information regarding this.

No, VLC does not support this. This is one of a number of reasons why we say Syncplay works best with mpv.

> 
> (side-note, chatting on mpv on Linux only lets me use ASCII characters and not the special characters of portuguese like ã, ó, etc, but I can do so inside the syncplay window, does anyone know why?)

The Syncplay's  mpv Lua script only recognises certain keyboard inputs and they have to be hard coded. You can try adding them to Syncplayintf.lua as per  https://github.com/Syncplay/syncplay/commit/aa5e7d8965f8ee4e97d60a54afb4cf5ffe5c8b9e#diff-56ee60d65a2cf44bf413a63b22e9929b3eee302372fb67444e2e600ff7be9098 and if it works then make a pull request or post all the characters here.

**Comment by jvschiavo** [comment]
Thanks for the quick and helpful answer!

I'm kinda curious as to what mpv has that VLC doesn't have that makes this possible. Is there any other player with this capability? Maybe could we add this to the VLC code?

Could we just add all the Unicode characters into that "non_us_chars" list? Or doing so would have any disadvantage? Because the benefit is instant support for all characters in languages (and even emoji I guess :P).

**Comment by Et0h** [comment]
> Thanks for the quick and helpful answer!
> 
> I'm kinda curious as to what mpv has that VLC doesn't have that makes this possible. Is there any other player with this capability? Maybe could we add this to the VLC code?

If you look at https://github.com/Syncplay/syncplay/blob/master/syncplay/resources/syncplayintf.lua you will see that to get in-player chat working in mpv we:
(a) Have dynamic key binding to capture the input keys and temporarily supress any other bindings they might be associated with; and
(b) Display input as it is being typed using a custom OSD layer, with text wrapping.

My understanding is that the only media player which supports (a) is mpv. If anyone wants to try getting that to work in VLC then feel free.

> 
> Could we just add all the Unicode characters into that "non_us_chars" list? Or doing so would have any disadvantage? Because the benefit is instant support for all characters in languages (and even emoji I guess :P).

I don't think that is possible within mpv at present, but you could ask an mpv developer to see determine there is a sensible way to do it.

It has been stated that "as of Version 13.0, the Unicode Standard contains 143,859 characters" - I don't think mpv would be happy if we tried binding them all. Also, who has an emoji keyboard? If people want emojis then they can paste them into the Syncplay chat input (or into mpv using ctrl+v while the input mode is on, but it does cause a slight pause so is not ideal).

For now our approach is that if people find characters they want to use but cannot they can submit a request to add one, specifying the precise characters in all desired cases.

**Comment by jvschiavo** [comment]
The only characters that I can think of that we need and might be missing are those:
ç ñ
´ á é í ó ú
` à è ì ò ù
^ â ê î ô û
~ ã ẽ ĩ õ ũ
¨ ä ë ï ö ü

By the way, a friend of mine realized some of those characters are already in the script, but they don't work because it is missing support for something like "dead keys". She said that in most languages we use dead keys to input some characters, and it is not working, so we can't input most of them.

**Comment by soredake** [comment]
Any progress on this?

**Comment by Et0h** [comment]
Okay, I've added to characters and then deleted the suspected duplicates - if they don't work then that is probably an mpv or OS issue rather than a Syncplay issue. There might be some workaround in terms of detecting ctrl/shift plus a different character, but I don't use the relevant keyboards so would not be able to implement such a workaround. As such, I'm closing this issue as I've got as far as I can on it but if someone wants to make a pull request with further improvements then feel free.

**Comment by AtilioA** [comment]
Neither 3e03747 nor d3f2c43 work for me on Windows 10. Previous behavior was no special characters appeared at all; now they appear as soon as they are pressed (and not compounded to the letters).
Update: Didn't work on Linux either.

**Comment by AtilioA** [comment]
@Et0h shouldn't this issue remain open if someone wants to tackle this, as the problem still persists? In fact, it's probably better to open another issue since this problem is just a side note to the current issue, which should be closed; I could open the new issue. I've messed with the `bindings` variable but I could not bind '`´+e`' to produce `é`, for example.

**Comment by Et0h** [comment]
> @Et0h shouldn't this issue remain open if someone wants to tackle this, as the problem still persists?

Suggested enhancement aren't really 'issues'. However, I can put it over to a 'discussion' to see if anyone wants to give it a try.

**Reply by AtilioA** [reply]
Well, it is an issue, I can't type diacritics on mpv.

**Reply by Et0h** [reply]
> Well, it is an issue, I can't type diacritics on mpv.

As far as I'm aware that's an mpv issue rather than a Syncplay issue.

**Reply by AtilioA** [reply]
In my understanding, the problem lies in the syncplayintf.lua file, since it is possible to place special characters in the input and for them to be displayed, but not to compose them into letters as ´+e = é. However, I have no knowledge of Lua, so the problem might really be out of Syncplay's reach.
